stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,376 @@
+                // TODO: must check if the first block was poorly timed.
+                warn!(
+                    "Miner is not building off of most recent tenure, but a tenure they attempted to reorg has already mined blocks.";
+                    "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                    "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                    "parent_tenure" => %sortition_state.parent_tenure_id,
+                    "last_sortition" => %sortition_state.prior_sortition,
+                    "violating_tenure_id" => %tenure.consensus_hash,
+                    "violating_tenure_first_block_id" => ?tenure.first_block_mined,
+                );
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+
+    fn check_tenure_change_block_confirmation(
+        tenure_change: &TenureChangePayload,
+        block: &NakamotoBlock,
+        signer_db: &SignerDb,
+    ) -> Result<bool, ClientError> {
+        // in tenure changes, we need to check:
+        // (1) if the tenure change confirms the expected parent block (i.e.,
+        //     the last block we signed in the parent tenure)
+        // (2) if the parent tenure was a valid choice
+        Self::confirms_known_blocks_in(block, &tenure_change.prev_tenure_consensus_hash, signer_db)
+    }
+
+    fn confirms_known_blocks_in(
+        block: &NakamotoBlock,
+        tenure: &ConsensusHash,
+        signer_db: &SignerDb,
+    ) -> Result<bool, ClientError> {
+        let Some(last_known_block) = signer_db
+            .get_last_signed_block_in_tenure(tenure)
+            .map_err(|e| ClientError::InvalidResponse(e.to_string()))?
+        else {
+            info!(
+                "Have not signed off on any blocks in the parent tenure, assuming block confirmation is correct";
+                "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                "tenure" => %tenure,
+            );
+            return Ok(true);
+        };
+        if block.header.chain_length > last_known_block.block.header.chain_length {
+            Ok(true)
+        } else {
+            warn!(
+                "Miner block proposal's tenure change transaction does not confirm as many blocks as we expect in the parent tenure";
+                "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                "proposed_chain_length" => block.header.chain_length,
+                "expected_at_least" => last_known_block.block.header.chain_length + 1,
+            );
+            Ok(false)
+        }
+    }
+
+    /// Has the current tenure lasted long enough to extend the block limit?
+    pub fn tenure_time_passed_block_lim() -> Result<bool, ClientError> {
+        // TODO
```

Comment:
Is there an issue for this?

Reply:
Added https://github.com/stacks-network/stacks-core/issues/4843